receiver/factorytemplate/factory_template.go

```diff
@@ -0,0 +1,195 @@
+type metricsReceiverFactory struct {
+	factory
+	newReceiver func(interface{}, consumer.MetricsConsumer) (receiver.MetricsReceiver, error)
+}
+
+var _ (receiver.MetricsReceiverFactory) = (*metricsReceiverFactory)(nil)
+
+// NewTraceReceiverFactory creates a factory for the given receiver "type" that
+// will have as the default configuration the object returned by newDefaultCfg and
+// that can be created using newReceiver from the configuration type returned by
+// newDefaultCfg.
+//
+// The object returned by newDefaultCfg should have its fields properly decorated using
+// the mapstructure attribute.
+//
+// The first parameter passed to newReceiver is going to be from the same type returned by
+// newDefaultCfg. The receiver implementer is going to be able to cast it to the appropriate
+// type.
+func NewTraceReceiverFactory(
+	receiverType string,
+	newDefaulfCfg func() interface{},
+	newReceiver func(interface{}, consumer.TraceConsumer) (receiver.TraceReceiver, error),
+) (receiver.TraceReceiverFactory, error) {
+	if receiverType == "" {
+		return nil, ErrEmptyReciverType
+	}
+	if newDefaulfCfg == nil {
+		return nil, ErrNilNewDefaultCfg
+	}
+	if newReceiver == nil {
+		return nil, ErrNilNewReceiver
+	}
+
+	return &traceReceiverFactory{
+		factory: factory{
+			receiverType:  receiverType,
+			newDefaultCfg: newDefaulfCfg,
+		},
+		newReceiver: newReceiver,
+	}, nil
+}
+
+// NewMetricsReceiverFactory creates a factory for the given receiver "type" that
+// will have as the default configuration the object returned by newDefaultCfg and
+// that can be created using newReceiver from the configuration type returned by
+// newDefaultCfg.
+//
+// The object returned by newDefaultCfg should have its fields properly decorated using
+// the mapstructure attribute.
+//
+// The first parameter passed to newReceiver is going to be from the same type returned by
+// newDefaultCfg. The receiver implementer is going to be able to cast it to the appropriate
+// type.
+func NewMetricsReceiverFactory(
+	receiverType string,
+	newDefaulfCfg func() interface{},
+	newReceiver func(interface{}, consumer.MetricsConsumer) (receiver.MetricsReceiver, error),
+) (receiver.MetricsReceiverFactory, error) {
+	if receiverType == "" {
+		return nil, ErrEmptyReciverType
+	}
+	if newDefaulfCfg == nil {
+		return nil, ErrNilNewDefaultCfg
+	}
+	if newReceiver == nil {
+		return nil, ErrNilNewReceiver
+	}
+
+	return &metricsReceiverFactory{
+		factory: factory{
+			receiverType:  receiverType,
+			newDefaultCfg: newDefaulfCfg,
+		},
+		newReceiver: newReceiver,
+	}, nil
+}
+
+// Type gets the type of the receiver created by this factory.
+func (f *factory) Type() string {
+	return f.receiverType
+}
+
+// NewFromViper takes a viper.Viper configuration and creates a new TraceReceiver.
+// Returning also the configuration used to create it.
+func (trf *traceReceiverFactory) NewFromViper(v *viper.Viper, next consumer.TraceConsumer) (receiver.TraceReceiver, interface{}, error) {
```

Comment:
The name newfromviper may cause a problem for OC which tries to implement both trace/metrics in the same struct. The requirement to be in the same struct comes from the fact that we need to share the grpc server (same addr:port).

How do we plan to handle that?

Reply:
We have 2 different interfaces: TraceReceiverFactory and MetricsReceiverFactory, the NewFromViper on each interface returns a different type, respectively TraceReceiver and MetricsReciver so OC receiver has both factories. 

Receivers that handle both trace and metrics on a single endpoint need some code to handle if the same "server/endpoint" is handling both, e.g.: they need to be a singleton for a given endpoint and internally it should use the same for metrics and traces.